File: emsclaw/backend/emsclaw_backend/entity/pcs.py
"""PCS / Battery Entity — 跨域数据契约(全局 DTO,前后端共享枚举)。"""
from __future__ import annotations
from enum import Enum
from typing import Optional

from pydantic import BaseModel, ConfigDict


class PcsMode(str, Enum):
    CHARGE = "charge"
    DISCHARGE = "discharge"
    STANDBY = "standby"
    AUTO = "auto"


class PcsStatus(str, Enum):
    RUNNING = "running"
    STANDBY = "standby"
    FAULT = "fault"


class PcsDeviceDTO(BaseModel):
    """PCS 配置(对应 ORM PcsDevice)。"""
    model_config = ConfigDict(from_attributes=True)
    id: str
    device_id: str
    rated_power_kw: float
    rated_reactive_kvar: float
    ac_voltage_v: float
    dc_voltage_v: float
    rated_efficiency: float
    min_soc: float
    max_soc: float
    override_mode: Optional[str] = None
    override_power_kw: Optional[float] = None
    override_expires_at: int = 0
    battery_id: Optional[str] = None
    created_at: int = 0
    updated_at: int = 0


class BatteryDeviceDTO(BaseModel):
    """电池配置(对应 ORM BatteryDevice)。"""
    model_config = ConfigDict(from_attributes=True)
    id: str
    device_id: str
    rated_capacity_kwh: float
    rated_voltage_v: float
    rated_current_a: float
    soc: float
    soh: float
    cycle_count: int
    created_at: int = 0
    updated_at: int = 0


class PcsSnapshotDTO(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    pcs_id: str
    timestamp: int
    active_power_kw: float
    reactive_power_kvar: float
    mode: str
    ac_voltage_v: float
    dc_voltage_v: float
    efficiency: float
    status: str


class BatterySnapshotDTO(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    battery_id: str
    timestamp: int
    soc: float
    voltage: float
    current_a: float
    temperature: float
    mode: str
    cycle_count: int

File: emsclaw/frontend/src/types/pcs.ts
// 与 emsclaw_backend/entity/pcs.py 对应
export type PcsMode = 'charge' | 'discharge' | 'standby' | 'auto';

export interface PcsDevice {
  id: string;
  device_id: string;
  rated_power_kw: number;
  rated_reactive_kvar: number;
  ac_voltage_v: number;
  dc_voltage_v: number;
  rated_efficiency: number;
  min_soc: number;
  max_soc: number;
  override_mode: PcsMode | null;
  override_power_kw: number | null;
  override_expires_at: number;
  battery_id: string | null;
  created_at: number;
  updated_at: number;
}

export interface BatteryDevice {
  id: string;
  device_id: string;
  rated_capacity_kwh: number;
  rated_voltage_v: number;
  rated_current_a: number;
  soc: number;
  soh: number;
  cycle_count: number;
  created_at: number;
  updated_at: number;
}

export interface PcsSnapshot {
  id: string;
  pcs_id: string;
  timestamp: number;
  active_power_kw: number;
  reactive_power_kvar: number;
  mode: PcsMode;
  ac_voltage_v: number;
  dc_voltage_v: number;
  efficiency: number;
  status: string;
}

export interface BatterySnapshot {
  id: string;
  battery_id: string;
  timestamp: number;
  soc: number;
  voltage: number;
  current_a: number;
  temperature: number;
  mode: PcsMode;
  cycle_count: number;
}

export interface PcsStatusItem {
  pcs: PcsDevice;
  battery: BatteryDevice | null;
  latest_pcs_snapshot: PcsSnapshot | null;
  latest_battery_snapshot: BatterySnapshot | null;
}

export const PCS_MODE_LABEL: Record<PcsMode, string> = {
  charge: '充电',
  discharge: '放电',
  standby: '待机',
  auto: '自动',
};

// 今日充放电调度计划(对应 PcsService.get_charge_schedule)
export interface ChargeScheduleSegment {
  start_hour: number;
  end_hour: number;
  start: string;   // "HH:00"
  end: string;     // "HH:00"
  mode: PcsMode;
  label: string;   // 充电(低谷) / 放电(高峰) / 放电(尖峰) / 待机
  power_ratio_lo: number;
  power_ratio_hi: number;
  power_kw: number; // 按在线 PCS 总额定功率 × ratio 上界换算的展示功率
}

export interface ChargeSchedule {
  date: string; // "YYYY-MM-DD"
  total_rated_power_kw: number;
  strategies: string[] | null;
  status: string;
  schedule_id: string | null;
  created_by: string | null; // session_id(=thread_id),可用于跳转到创建该计划的会话
  segments: ChargeScheduleSegment[];
}
